nomad/state/filter_test.go

```diff
@@ -0,0 +1,157 @@
+package state
+
+import (
+	"testing"
+	"time"
+
+	"github.com/hashicorp/go-bexpr"
+	"github.com/hashicorp/nomad/helper/uuid"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+func BenchmarkEvalListFilter(b *testing.B) {
```

Comment:
Benchmark results from my machine:

```
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkEvalListFilter/filter_with_index-16                3700           6832056 ns/op             458 B/op         13 allocs/op
BenchmarkEvalListFilter/filter_with_go-bexpr-16              100         234049608 ns/op        47202395 B/op    4000009 allocs/op
BenchmarkEvalListFilter/paginated_filter_with_index-16           4819960              4628 ns/op            2616 B/op         24 allocs/op
BenchmarkEvalListFilter/paginated_filter_with_go-bexpr-16          59338            409490 ns/op          122438 B/op       8325 allocs/op
```

Filtered results are an order or two of magnitude slower than using an index so we should document query params that are index based. 

I also had a version that used [`FilterIterator`](https://pkg.go.dev/github.com/hashicorp/go-memdb#FilterIterator), but the result was the same as `filter_with_go-bexpr`

Reply:
The 100 eval page in the paginated results are definitely more "realistic" than us trying to iterate over 100k evals at once. 

Something else to keep in mind looking at this is that in `paginated_filter_with_go-bexpr` we're creating the bexpr evaluator every iteration whereas in `filter_with_go-bexpr` we're creating the bexpr evaluator once outside the `b.N` loop. Unfortunately a given API call doesn't get to amortize creating the evaluator across multiple pages, so `paginated_filter_with_go-bexpr` is going to be even worse.

Reply:
Ah good point, should I move the call to `CreateEvaluator` so it's within the bechmark `for` loop?

Another thing that I thought is that there's a diminishing advantage of using pagination when querying later pages since we need to iterate over everything that comes before anyway, even if we don't apply the filter:

```
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkEvalListFilter/filter_with_index-16                3673           6704301 ns/op
BenchmarkEvalListFilter/filter_with_go-bexpr-16              100         220508127 ns/op
BenchmarkEvalListFilter/paginated_filter_with_index-16           5146934              4724 ns/op
BenchmarkEvalListFilter/paginated_filter_with_go-bexpr-16          55220            480332 ns/op
BenchmarkEvalListFilter/paginated_filter_with_index_last_page-16                    2017          11966223 ns/op
BenchmarkEvalListFilter/paginated_filter_with_go-bexpr_last_page-16                  872          27409941 ns/op
```